jaeger-core/src/main/java/io/jaegertracing/Configuration.java

```diff
@@ -14,27 +14,27 @@
 
 package io.jaegertracing;
 
-import io.jaegertracing.metrics.Metrics;
-import io.jaegertracing.metrics.MetricsFactory;
-import io.jaegertracing.metrics.NoopMetricsFactory;
-import io.jaegertracing.propagation.B3TextMapCodec;
-import io.jaegertracing.propagation.Codec;
-import io.jaegertracing.propagation.CompositeCodec;
-import io.jaegertracing.propagation.TextMapCodec;
-import io.jaegertracing.reporters.CompositeReporter;
-import io.jaegertracing.reporters.LoggingReporter;
-import io.jaegertracing.reporters.RemoteReporter;
-import io.jaegertracing.reporters.Reporter;
-import io.jaegertracing.samplers.ConstSampler;
-import io.jaegertracing.samplers.HttpSamplingManager;
-import io.jaegertracing.samplers.ProbabilisticSampler;
-import io.jaegertracing.samplers.RateLimitingSampler;
-import io.jaegertracing.samplers.RemoteControlledSampler;
-import io.jaegertracing.samplers.Sampler;
-import io.jaegertracing.senders.Sender;
-import io.jaegertracing.senders.SenderFactory;
-import io.jaegertracing.senders.SenderResolver;
-import io.opentracing.Tracer;
+import io.jaegertracing.internal.JaegerTracer;
+import io.jaegertracing.internal.metrics.Metrics;
+import io.jaegertracing.internal.metrics.MetricsFactory;
```

Comment:
MetricsFactory and NoopMetricsFactory are part of SPI

Reply:
I actually left the `MetricsFactory` out of the SPI on purpose, as I thought this was too internal to be customized. Given that we already delegate to an abstraction (Micrometer), it means people can just build a Micrometer reporter instead.

About `NoopMetricsFactory`, I don't think it belongs to the SPI. Would you mind sharing why you think it would?

Reply:
You're probably right about noop factory, it's just used as a default. But I think MetricsFactory belongs to an SPI because we don't know upfront which metrics API people want to use with the  client. Does Micrometer really support _everything_?

Reply:
It's an abstraction with tons of providers already implemented: http://micrometer.io/docs

If a provider is not there, people would either have to implement a micrometer provider or a `MetricsFactory` provider, so, pretty much the same work (but their Micrometer might be reused by other applications and in other scenarios). 

I'd really encourage our users to use Micrometer with their own provider than implementing their own `MetricsFactory` provider.

Reply:
That's assuming people use oss frameworks. For example, we internally have a Java client for Uber's m3db metrics system. Rather than implementing an adapter through Micrometer, it makes more sense for us to bind directly to jaeger's MetricsFactory (which we already do).

Reply:
> That's assuming people use oss frameworks

In general, like, Jaeger? :-) Or do you mean for the metrics backend, like Prometheus?

> Rather than implementing an adapter through Micrometer, it makes more sense for us to bind directly to jaeger's MetricsFactory (which we already do).

I would argue that this is like using the (internal) `JaegerTracer.Builder`: there are valid use cases, but it's not the usual case. The usual case is for an application to be using a framework like Micrometer to make data available to a backend like Prometheus. 

How about we keep this internal first and, if there's demand, we make it part of the SPI?

Reply:
> In general, like, Jaeger? :-) Or do you mean for the metrics backend, like Prometheus?

I mean oss frameworks for reporting metrics. Organizations adopting Jaeger may be running all kinds of custom metrics backends, respectively with custom libraries for sending metrics to those backends, which is the case at Uber, as I don't think our metrics client for Java is even open source.

I am not suggesting making MetricsFactory a public interface, but moving it to `spi`. You already have a bunch of interfaces there, which in my view are actually a lot less likely to be implemented by other people, e.g. who would need to implement Sampler or Sender?  If people do implement those, they are doing something very tracing-specific, whereas implement MetricsFactory is simply adapting Jaeger lib to run in a custom enterprise environment.